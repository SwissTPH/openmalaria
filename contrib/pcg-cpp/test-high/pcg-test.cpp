/*
 * PCG Random Number Generation for C++
 *
 * Copyright 2014-2017 Melissa O'Neill <oneill@pcg-random.org>,
 *                     and the PCG Project contributors.
 *
 * SPDX-License-Identifier: (Apache-2.0 OR MIT)
 *
 * Licensed under the Apache License, Version 2.0 (provided in
 * LICENSE-APACHE.txt and at http://www.apache.org/licenses/LICENSE-2.0)
 * or under the MIT license (provided in LICENSE-MIT.txt and at
 * http://opensource.org/licenses/MIT), at your option. This file may not
 * be copied, modified, or distributed except according to those terms.
 *
 * Distributed on an "AS IS" BASIS, WITHOUT WARRANTY OF ANY KIND, either
 * express or implied.  See your chosen license for details.
 *
 * For additional information about the PCG random number generation scheme,
 * visit http://www.pcg-random.org/.
 */

/*
 * This file is based on the demo program for the C generation schemes.
 * It shows some basic generation tasks.
 */

#include <cstddef>
#include <cstdint>
#include <cstdlib>
#include <cassert>
#include <climits>
#include <iostream>
#include <sstream>
#include <iomanip>
#include <algorithm>
#include <numeric>
#include <random>       // for random_device

#include "pcg_random.hpp"

// This code can be compiled with the preprocessor symbol RNG set to the
// PCG generator you'd like to test.

#ifndef RNG
    #define RNG pcg32
    #define TWO_ARG_INIT 1
#endif

#define STRINGIFY_IMPL(x) #x
#define STRINGIFY(x)      STRINGIFY_IMPL(x)

using namespace std;
using pcg_extras::operator<<;

#if !PCG_EMULATED_128BIT_MATH || !AWKWARD_128BIT_CODE

int main(int argc, char** argv)
{
    // Read command-line options

    int rounds = 5;
    bool nondeterministic_seed = false;

    ++argv;
    --argc;
    if (argc > 0 && strcmp(argv[0], "-r") == 0) {
        nondeterministic_seed = true;
        ++argv;
        --argc;
    }
    if (argc > 0) {
        rounds = atoi(argv[0]);
    }

    /* Many of the generators can be initialized with two arguments; the second
     * one specifies the stream.
     */

#if TWO_ARG_INIT
    RNG rng(42u, 54u);
#else
    RNG rng(42u);
#endif

    if (nondeterministic_seed) {
        // Seed with external entropy from std::random_device (a handy
        // utility provided by pcg_extras).
        rng.seed(pcg_extras::seed_seq_from<std::random_device>());
    }

    constexpr auto bits = sizeof(RNG::result_type) * CHAR_BIT;
    constexpr int how_many_nums = bits <= 8  ? 14
                                : bits <= 16 ? 10
                                :              6;
    constexpr int wrap_nums_at  = bits >  64 ? 2
                                : bits >  32 ? 3
                                :              how_many_nums;

    cout << STRINGIFY(RNG) << ":\n"
    //   << "      -  aka:         " << pcg_extras::printable_typename<RNG>()
    // ^-- we skip this line because the output is long, scary, ugly, and
    //     and varies depending on the platform
         << "      -  result:      " <<  bits << "-bit unsigned int\n"
         << "      -  period:      2^" << RNG::period_pow2();
    if (RNG::streams_pow2() > 0)
         cout << "   (* 2^" << RNG::streams_pow2() << " streams)";
    cout << "\n      -  size:        " << sizeof(RNG) << " bytes\n\n";

    for (int round = 1; round <= rounds; ++round) {
        printf("Round %d:\n", round);

        /* Make some N-bit numbers */
        cout << setw(4) << setfill(' ') << bits << "bit:";
        for (int i = 0; i < how_many_nums; ++i) {
            if (i > 0 && i % wrap_nums_at == 0)
                cout << "\n\t";
            cout << " 0x" << hex << setfill('0') 
                 << setw(sizeof(RNG::result_type)*2) << rng();
        }
        cout << endl;

        cout << "  Again:";
        rng.backstep(6);
        for (int i = 0; i < how_many_nums; ++i) {
            if (i > 0 && i % wrap_nums_at == 0)
                cout << "\n\t";
            cout << " 0x" << hex << setfill('0') 
                 << setw(sizeof(RNG::result_type)*2) << rng();
        }
        cout << dec << endl;

        /* Toss some coins */
        cout << "  Coins: ";
        for (int i = 0; i < 65; ++i)
            cout << (rng(2) ? "H" : "T");
        cout << endl;

        RNG rng_copy{rng};
        /* Roll some dice */
        printf("  Rolls:");
        for (int i = 0; i < 33; ++i)
            cout << " " << (uint32_t(rng(6)) + 1);
        cout << "\n   -->   rolling dice used "
             << static_cast<size_t>(rng - rng_copy)
             << " random numbers" << endl;

        /* Deal some cards using pcg_extras::shuffle, which follows
         * the algorithm for shuffling that most programmers would expect.
         * (It's unspecified how std::shuffle works.)
         */      
        enum { SUITS = 4, NUMBERS = 13, CARDS = 52 };
        char cards[CARDS];
        iota(begin(cards), end(cards), 0);
        pcg_extras::shuffle(begin(cards), end(cards), rng);

        /* Output the shuffled deck */
        printf("  Cards:");
        static const signed char number[] = {'A', '2', '3', '4', '5', '6', '7',
                                             '8', '9', 'T', 'J', 'Q', 'K'};
        static const signed char suit[] =   {'h', 'c', 'd', 's'};
        int i = 0;
        for (auto card : cards) {
            ++i;
            cout << " " << number[card / SUITS] << suit[card % SUITS];
            if (i % 22 == 0)
                cout << "\n\t";
        }
        
        cout << "\n" << endl;
    }

    // Ensure that the input/output functions work.
    std::stringstream ss;
    ss << rng;
    RNG from_string;
    ss >> from_string;
    assert(from_string == rng);

    return 0;
}

#else //  i.e. PCG_EMULATED_128BIT_MATH && AWKWARD_128BIT_CODE

int main()
{
    // It's not that it *can't* be done, it just requires either C++14-style
    // constexpr or some things not to be declared const.
    cout << "Sorry, " STRINGIFY(RNG) " not supported with emulated 128-bit math"
         << endl;
}

#endif
