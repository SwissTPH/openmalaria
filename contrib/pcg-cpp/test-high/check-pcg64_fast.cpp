#define RNG pcg64_fast
#define TWO_ARG_INIT 0

#include "pcg-test.cpp"

