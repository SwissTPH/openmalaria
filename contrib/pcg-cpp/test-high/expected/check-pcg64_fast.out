pcg64_fast:
      -  result:      64-bit unsigned int
      -  period:      2^126
      -  size:        16 bytes

Round 1:
  64bit: 0x63b4a3a813ce700a 0x382954200617ab24 0xa7fd85ae3fe950ce
	 0xd715286aa2887737 0x60c92fee2e59f32c 0x84c4e96beff30017
  Again: 0x63b4a3a813ce700a 0x382954200617ab24 0xa7fd85ae3fe950ce
	 0xd715286aa2887737 0x60c92fee2e59f32c 0x84c4e96beff30017
  Coins: HTTTTTTTTTTTHHTHHHHHHTTHTHTTHTTTTTHTTHTTHHHHTHHTHHHHHTTTHHHHHHHHH
  Rolls: 4 5 5 4 4 4 5 4 3 1 6 1 6 6 2 6 2 2 3 3 2 5 6 4 2 6 4 4 3 2 4 2 6
   -->   rolling dice used 33 random numbers
  Cards: Jh Jc 2h 4h Kd 2c 4d 5c Kc 8c 7h Td 9s 4s 2s Jd 8s 6h Qd 6d 9c Qh
	 Tc 3h 7c 2d 4c 3s Qc 5s 8d 9d Ah Ac 3c 6s 5h 7s Ks Th 5d Ad Js Kh
	 9h Qs 6c 3d 7d 8h As Ts

Round 2:
  64bit: 0x824eb71d0f02dfb4 0x7aaad63730e335c1 0xf87271e197a74023
	 0x86d11317e615a346 0xe067147998450163 0x59fc13368ae72993
  Again: 0x824eb71d0f02dfb4 0x7aaad63730e335c1 0xf87271e197a74023
	 0x86d11317e615a346 0xe067147998450163 0x59fc13368ae72993
  Coins: TTHHTHTHTTTHTTTHHHTTTHTTHTHHHTTHHTHHTHHTHTTHTTHTHHHTTTTHTHTTTTTHH
  Rolls: 1 5 2 5 1 5 5 3 3 3 5 6 3 5 2 4 1 2 1 4 6 2 5 2 2 4 3 6 5 1 2 5 5
   -->   rolling dice used 33 random numbers
  Cards: Th Qs 2c 4d Jc Jd 4c 2h 4s Tc Kh 6d 3c 7s 5c 6c 6h Kc Qh 6s 2d 4h
	 9s 8c 3s 9c 7c 8d Kd 7d Ac 3h Qc 5s 5h As 8h Ks Ts 8s 9d Qd Js Td
	 Jh 2s Ad 7h 9h Ah 5d 3d

Round 3:
  64bit: 0xaf64b2a730ffd1b9 0xef9f2e946e08fbe3 0x181b81b6a0b5bee2
	 0x219851c742250cf4 0xb459875e221e7df5 0xe7518dd5d411bae8
  Again: 0xaf64b2a730ffd1b9 0xef9f2e946e08fbe3 0x181b81b6a0b5bee2
	 0x219851c742250cf4 0xb459875e221e7df5 0xe7518dd5d411bae8
  Coins: HTHHTHTTTHHTTHHTTTTHTTHHTHTTHHHTTHTTHHHTTTHTTTTHHHHHTTTTTHHTHTTTT
  Rolls: 1 3 6 6 2 1 1 4 2 5 1 3 1 5 1 5 1 6 2 3 1 2 3 4 2 4 5 4 2 2 4 5 2
   -->   rolling dice used 33 random numbers
  Cards: Jc Jh 2s 3s Ac 5s Td Kc 7s Th 7d 5h 3d As Qd 4h 3h Js 4c 5d Tc 4d
	 8s Ah 6h 5c 2c 7c 3c Qs 6s 2h Ks Ts 9c 2d Kd 6c 8h Qh 9s Ad 9d 4s
	 Jd 8c 7h 6d Qc 8d Kh 9h

Round 4:
  64bit: 0xe7e1d1711485e473 0xc0c6d1f0e72a55d3 0x2d0fa33eb3638524
	 0xd0cb8d73a16deacd 0x3e410a9cc7682918 0x8df7d57b4d2f9ac4
  Again: 0xe7e1d1711485e473 0xc0c6d1f0e72a55d3 0x2d0fa33eb3638524
	 0xd0cb8d73a16deacd 0x3e410a9cc7682918 0x8df7d57b4d2f9ac4
  Coins: THHTHTHTHHTHTHHTTHTHHTHHHTTTHTTHHHHTTTTHTTHTTHTTTTHTHHHTHTTHTHTTH
  Rolls: 5 4 3 2 6 2 6 6 2 3 1 3 1 6 2 4 3 6 6 6 4 1 4 6 5 1 3 6 3 5 1 6 1
   -->   rolling dice used 33 random numbers
  Cards: 3d 4h 2d 8d Js Qs 3s 4d 9h Ks Tc Qc Qh 9c 7s 6s 6c Th 6h Ts 9s Jh
	 2s 7c Td Ac Jc 8c 6d 4s 9d As 7d 5h 3h 2h Ad 4c 8s 8h 5s Kh Ah Kc
	 3c Jd 5c 7h Kd Qd 5d 2c

Round 5:
  64bit: 0x2a2d820a8f859ba2 0x1a74e59c8e288526 0x6b856b08000af65c
	 0x793c0d4103ce2a55 0xc5081bea922d1d0c 0x6b61da59d73efbf9
  Again: 0x2a2d820a8f859ba2 0x1a74e59c8e288526 0x6b856b08000af65c
	 0x793c0d4103ce2a55 0xc5081bea922d1d0c 0x6b61da59d73efbf9
  Coins: HTTHTHTTHTTTTTHTHTHHTHTHHHTHHHTTHHTHTTTTTHHHHHTHHHTTTHHTTTTHHHHHH
  Rolls: 4 4 1 3 5 3 1 3 4 3 1 4 2 5 6 2 2 5 6 3 4 2 4 5 6 6 4 2 4 3 1 5 1
   -->   rolling dice used 33 random numbers
  Cards: Ks Ah Ac Ad 3d 9h 7c 3c 7d 3s Th 2d Jh 7h 5h Td 8d 6c Kd 2h Jc 9d
	 8s 2c 8h 4h Qd 6d 5c Js 5s 9c As 4s Jd Kh 2s Tc 8c 5d Qh 3h Qc 6s
	 Kc 7s 4d Ts Qs 6h 4c 9s

